### ISSUE TITLE
Separation of bounded (and repeatable) and unbounded (or not repeatable) Dataset

### ISSUE THREAD

**Issue by yongtang** [tensorflow/io#513]
For tensorflow-io we always start out IO related implementation with Dataset. The reason is to tie to TF 2.0 overall and tf.keras. In the past we only have limited input sources covered and everything was fine. However, as different input sources are exposed some issues arise.

While Dataset has been well supported in tf.keras, the mechanism of inference and training are different and that caused confusions and issues when a standard Dataset was passed to tf.keras and some functionalities stop working. To summarize:
1) Dataset is an iterable and could be unbounded (no EOF) and non-repeatable (re-run does not have to get the same result).
2) Inference could take any Dataset as input
3) Training could take a Dataset that is iterable. However, training also requires that
    a) The Dataset has to be bounded (with an EOF)
    b) The Dataset has to be repeatable (re-run will get the same result), as the Dataset node will be recreated at each epoch.

The issue was observed when Kafka dataset was passed and starting offset was not set. In that case re-run of Dataset causes data inconsistency across different epoch runs. However, Kafka itself is repeatable as long as the offset has been provided (as Kafka has an excellent offset management). In fact Kafka is even indexable and you can build even SQL on top of it. 

One counter example of Dataset is to read data from standard input (or network packet). There is no way you can repeat (unless caching the data on disk). In that case reading standard input can only be used for inference.

There were lots of confusion about Dataset and users encounter many issues as far as I know.

Since tensorflow-io intends to cover many input sources, I think it is best for us to split out the concept of repeatable and bounded Dataset, from the non-repeatable or non-bounded Dataset.

I am thinking (in combination to top level API exposure) about the following approach:
1) Expose an tfio.IODataset which is reserved for repeatable and bounded Dataset, and let user know IODataset is safe to be used with training.
2) Expose an tfio.StreamIODataset which is used for Dataset that is not safe to pass to training. Examples include standard input, grpc stream, from video caputuring and audio recording devices, etc. We also should let user know StreamIODataset can only be used for inference or iterate with `for d in dataset`.

In addition, I think we could follow the similar API of factory and have:
```
tfio.IODataset.from_parquet
tfio.IODataset.from_kafka
tfio.IODataset.from_ffmpeg
```

Note: as was discussed, Kafka could be either training safe (when offset is provided), or inference only (in a continuous fashion). From that standpoint we should have
```
tfio.IODataset.from_kafka

tfio.StreamIODataset.from_kafka 
```
to cover different scenarios.

/cc @terrytangyuan @BryanCutler

**Comment by yongtang** [comment]
Also on `tfio.IOTensor`: it is different from Dataset in that IOTensor is indexable. You could access IOTensor randomly. In certain situations IOTensor might be cached to avoid expensive and frequent data access to disk. This could either be done at the C++ level, or python level (like `FFMpegAudioIOTensor`. So,
```
tfio.StreamIODataset # iterable, unbounded and/or non-repeatable
tfio.IODataset # iterable and repeatable
tfio.IOTensor # indexable and repeatable
```

**Comment by terrytangyuan** [comment]
Thanks for the detailed background. It makes perfect sense for me to split this into two types. 

Regarding the way we expose them in top-level APIs, the approach you suggested looks good to me. We just have to give informative and instructive error messages to users when they misuse different types of IO Datasets.

**Comment by yongtang** [comment]
Also some of the background in recent implementations on some C++ code:

1) *Will IODataset require different C++ implementation?*

   IODataset/IOStreamDataset are just a standard tf.data.Dataset class with a tag marked so that user knows they are for different purposes (training vs. inference). Any C++ Dataset implementation could be marked either IODataset or IOStreamDataset. The glitch is that user may spend day (or weeks as some experienced) trying to figure out why training is not working or never converge with some Dataset.

   If there is an existing Dataset implementation, it could be directly attached to `IODataet.from_...` and everything works as before.

   In short, IODataset just groups Dataset into two types: for training, or for inference.

2) *The recent C++ code addition seems to be different from a standard Dataset subgraph, why?*

   Recently, the C++ code has been added to use template (in C++) so that the subgraph creation and management are hided away. This is with the hope that new contributors could focus on getting data in and out of a storage and avoid spending weeks to first find out how Dataset's TF subgraph mechanism works.

   *It might be better now (to not spend too much time on subgraph), but overall the C++ code is still confusing...*

   We tried very hard to improve and will continue to improve. If there are better ways please comment or raise an issue.

3) *What exactly is IOTensor? How is IOTensor different from Dataset?*

   The problem IOTensor tries to solve, is for scenarios of random access data.

   Imagine you have a big audio/video clip of hundreds of GBs of data. You also have a subtitle that shows `time => text`. You are only interested in a small sub clip within `1:30:00~1:30:10` (10 seconds), because the related subtitle has a keyword. It is hard to get the sub clip from Dataset because you have to read from the beginning until `1:30:00`. The reason is that Dataset has been designed as an iterable, not able to indexing.  IOTensor serves the goal of random access/indexing/slicing: it has a `__getitem__` API exposed so you could take a small subset of the data. (/cc @ivelin as `video/audio <=> text` mapping might be something you are interested.)

4) *Some recently added IODataset creates a Dataset of numbers (1,2,3..) first, then call a map function to return the "true" Dataset asked, why extra steps?*

   When we work on Dataset, we notice that while Dataset is a subgraph in TF, it does not have the same level of API stability as a normal TF kernel ops (e.g., v1 vs. v2 changes). This is natural as a TF kernel ops is just one node while Dataset could be more than one node. However, as tensorflow-io is a different repo, we have to constantly catch up to any changes. That has been quite hard as some of the tests still fails (e.g., #431).

   But, tensorflow-io's focus is to getting data from storage to TF graph, not to work on graph intensively. That is why we create a sequence number Dataset first then call a map function to get to the data.

    A "sequence number Dataset" is a primitive Dataset that will always be reliable no matter how Dataset API changes in TF repo. We even don't care how it is created. Once this dataset is created, we just need to provide a C++ kernel op node (one node) to map a sequence number to the data we want to read. The C++ kernel op node we implement here only need to conform to standard TF C++ op which will be much more simplified. This should also lower the bar for community to contribute.

5) *That sounds great, but some Dataset (like streams) can only be read sequentially, what if the input sequence numbers are out of order?*

   The sequence number generated, is just to provide state information. It is up to the map function to decide if the number makes sense or not. If out of order sequence is passed then an error could simply be thrown out. Keep in mind the final exposed API is Dataset, not necessarily the intermediate kernel ops. (/cc @terrytangyuan @BryanCutler hopefully that may explain some of the changes I made recently).

6) *What about some Dataset like Avro/Parquet/Arrow where data are grouped into Blocks/RowGroups/Batches*

   This is about the "unit" of the sequence number. If your implementation can only read one "batch" at a time, you could return a batch of records for each number in `[1, 2, 3,...]`. The returned "batch of records" does not have to be the same size as you could use `unbatch()` conveniently. In that situation, the "unit" for each sequence number is a batch (variable size of records).

   Alternatively, you could change the sequence number Dataset to produce a sequence with uneven intervals, e.g., `[0, 15, 80]` where each interval matches the `Block/RowGroup/Batch` size. Inside the primitive kernel implementation the input is the record offset, not the Block/RowGroup/Batch index.

   Either way, with the above, we could focus on reading data from input source than subgraph management in Dataset. The above also helps in caching, as Block/RowGroup/Batch level reading are preferred for Avro/Parquet/Arrow like data formats. (/cc @BryanCutler as that might be related to Arrow/Parquet)

7) *I still don't follow IOTensor, is it different from normal Tensor?*

   The difference is that IOTensor will not touch any data until it will be used, even if you apply IOTensor operations on top of. We don't have a lot of IOTensor operations yet (contributions welcome), but take `window` as an example case:

   When you call `window()` on IOTensor (e.g. `t1`), the returned IOTensor (e.g.`t2`) does not read any data from your disk. Instead, `t2` will just attach a new function `window` after the old `t1`'s `read` function.

   When data is ready to be used, accessing `t2` will call `read(start, stop)`, then apply `window()` on top of the tensor returned from `read(start, stop)`.

   In short, IOTensor is just a chain of functions that will be applied to input source, one by one. (/cc @terrytangyuan in case you have any questions).

8) *What about resource, the recent addition of Dataset always starts with a resource (DT_RESOUCE) data type*

   `Resource` data type is a reference that is necessary because any of the above discussed items requires holding a reference. We also need the "reference" to persists in memory and "resource" data type satisfies. You can consider "resource" as a blob in memory, or a C++ pointer of any type.

   Alternatively, we also could use `Variant` data type, which could even be serialized/deserialized so not only it persist, it also could be saved. However, `Variant` does not play well with Dataset because Dataset's C++ occupies Variant for its internal use. In the past we have to get around this issue in a very ugly way `https://github.com/tensorflow/io/blob/f4f2f0994ff2e8d20310f9502edb89710945bddc/tensorflow_io/core/kernels/dataset_ops.h#L1112-L1136`

   We are also open to any new ideas that could improve and make a better C++ experience.

9) *Could Dataset and IOTensor be consolidate? We should only have one code base for C++*

   That is what we are doing when we introduce the new code. We have different abstractions in place. At the minimum, if you could implement a `Read(start, stop)` function in C++ then both Dataset and IOTensor could be implemented in less than tens of lines of python code.

   The `Read(start, stop)` is just an API format. If a data input source does not support random access, it could just throw out an error. In that case, Dataset will work while IOTensor will not be exposed.

   Here is the diagram of similarities:
   ```
   Dataset: [index(0), index(1), index(2), ...]
        =>
   Output: [read(index(0), ?), read(index(1), ?), read(index(2), ?), ...]

   IOTensor: [index(0), index(1)]
        =>
   Output: [read(index(0), index(1))]

   IOTensor: [index(1), index(2)]
        =>
   Output: [read(index(1), index(2))]
   ```

10) *Have all of the above been implemented? Some mentioned features might be missing?*

    We have to change many places as there are too many formats to work on, the building blocks are mostly there and high level API are largely stabilized.

    We also welcome contribution in new ideas of improving (not necessarily the same as mentioned) as well.

/cc @terrytangyuan @BryanCutler, hope the above explains some of the recent changes.

**Comment by yongtang** [comment]
Another reason for `Resource` and using sequence number dataset as "seed": it splits reference out of the Dataset operation itself. In this way, Resource (reference) persist while Dataset has been recreated (during each training epoch run). This guarantees the Dataset is repeatable.

The issue was originated from the idea of using Kafka for training. While Kafka was used for streaming originally, itself is a persistent storage with offset management. In theory you could just read the same data from Kafka with the same offset. In other words, it is repeatable.

When we passed the Kafka Dataset directly, we encountered issue and spend quite some time trying to figure out why training becomes unpredictable, until we found out that each training epoch run will rerun the node. In that case, if the offset is not absolute (e.g., RD_KAFKA_OFFSET_TAIL) then data will be different. It is understandable but a better documentation on Dataset's assumptions will definitely help.

With "Resource" + "sequence number seed Dataset", it allows us to resolve offset at the creation of Resource. This way, when final Dataset is created, the resource does not change (only sequence number restarted from 0) so we get the same repeatable result across multiple epoch runs.

**Comment by BryanCutler** [comment]
Thanks for all of the details @yongtang , I agree this is an important distinction to make that could save a lot of user from improper training. 

One thing to add though, from my understanding you could still train a model correctly from a streaming dataset as long as the stream can be repeated with the same sequence.Correct me if I'm wrong, but I believe for training over epochs Keras calls `tf.data.Dataset.repeat` that will construct a new iterator for that dataset each pass. If that iterator produces the same data each pass, then training will be correct.

Will it still be possible that certain `IOStreamDataset`s which meet these requirements could be used for training?

**Comment by yongtang** [comment]
Thanks @BryanCutler. When iterator is created at each epoch, my understanding is that the dataset node will be recreated with the same input parameters. So if a Dataset can reproduce the same sequence of data with the same input argument, then training will be fine. The StreamDataset we discussed here, are strictly limited to those Dataset that could not repeat with the same input argument. One example is the stream dataset that takes standard input (file descriptor). Even though we pass the same file descriptor (of stdin) when dataset is repeated, the data sequence will be different, as Dataset don't cache internally.

**Comment by yongtang** [comment]
@BryanCutler One good way to see if training is possible for stream dataset, is to try with multiple iterators (in 2.0/eager mode) with the same created dataset:
```
# create dataset only once, (iterate multiple times, see below).
dataset = ...

# 1st run, with the same dataset returned from above
v1 = [d for d in dataset]

# 2nd run, with the same dataset returned from above
v1 = [d for d in dataset]

assert v1 == v2
```

If `v1 == v2` then the dataset is safe to be used for training.
